http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/sam/SamServiceInterp.scala

```diff
@@ -0,0 +1,138 @@
+import org.broadinstitute.dsde.workbench.model.{UserInfo, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+import org.http4s.AuthScheme
+import org.http4s.Credentials.Token
+import org.typelevel.log4cats.StructuredLogger
+
+import java.util.UUID
+import scala.jdk.CollectionConverters._
+
+class SamServiceInterp[F[_]](apiClientProvider: SamApiClientProvider[F],
+                             cloudAuthTokenProvider: CloudAuthTokenProvider[F]
+)(implicit
+  F: Async[F],
+  metrics: OpenTelemetryMetrics[F],
+  logger: StructuredLogger[F]
+) extends SamService[F] {
+
+  private val saScopes = Seq(
+    "https://www.googleapis.com/auth/userinfo.email",
+    "https://www.googleapis.com/auth/userinfo.profile",
+    StorageScopes.DEVSTORAGE_READ_ONLY
+  )
+
+  // TODO: pattern for retries
+
+  override def getPetServiceAccount(userInfo: UserInfo, googleProject: GoogleProject)(implicit
+    ev: Ask[F, AppContext]
+  ): F[WorkbenchEmail] =
+    for {
+      ctx <- ev.ask
+      googleApi <- apiClientProvider.googleApi(userInfo.accessToken.token)
+      pet <- F.blocking(googleApi.getPetServiceAccount(googleProject.value)).adaptError { case e: ApiException =>
+        SamException.create("Error getting pet service account from Sam", e, ctx.traceId)
+      }
+    } yield WorkbenchEmail(pet)
+
+  override def getPetManagedIdentity(userInfo: UserInfo, azureCloudContext: AzureCloudContext)(implicit
+    ev: Ask[F, AppContext]
+  ): F[WorkbenchEmail] =
+    for {
+      ctx <- ev.ask
+      azureApi <- apiClientProvider.azureApi(userInfo.accessToken.token)
+      pet <- F
+        .blocking(
+          azureApi.getPetManagedIdentity(
+            new GetOrCreateManagedIdentityRequest()
+              .tenantId(azureCloudContext.tenantId.value)
+              .subscriptionId(azureCloudContext.subscriptionId.value)
+              .managedResourceGroupName(azureCloudContext.managedResourceGroupName.value)
+          )
+        )
+        .adaptError { case e: ApiException =>
+          SamException.create("Error getting pet managed identity from Sam", e, ctx.traceId)
+        }
+    } yield WorkbenchEmail(pet)
+
+  override def getProxyGroup(
+    userEmail: WorkbenchEmail
+  )(implicit ev: Ask[F, AppContext]): F[WorkbenchEmail] =
+    for {
+      ctx <- ev.ask
+      leoToken <- getLeoAuthToken
+      googleApi <- apiClientProvider.googleApi(leoToken)
+      proxy <- F.blocking(googleApi.getProxyGroup(userEmail.value)).adaptError { case e: ApiException =>
+        SamException.create("Error getting proxy group from Sam", e, ctx.traceId)
+      }
+    } yield WorkbenchEmail(proxy)
+
+  override def getPetServiceAccountToken(userEmail: WorkbenchEmail, googleProject: GoogleProject)(implicit
+    ev: Ask[F, AppContext]
+  ): F[String] = for {
+    ctx <- ev.ask
+    leoToken <- getLeoAuthToken
+    googleApi <- apiClientProvider.googleApi(leoToken)
+    // TODO: previously Leo cached this Sam call. Is that necessary?
```

Comment:
I am not sure, I could not find documentation around our logic to cache SAM calls like this. If Sam can handle the requests I would rather not cache

Reply:
Yeah that's my thinking too, I'd rather not cache. I'll remove this TODO (and do some testing)